### ISSUE TITLE
Elasticsearch-spark connector failing to save data with an Illegal Argument Exception : "No class name given"

### ISSUE THREAD

**Issue by eliasah** [elastic/elasticsearch-hadoop#837]
### What kind an issue is this?

- [x] Bug report. If you’ve found a bug, please provide a code snippet or test to reproduce it below.   
    The easier it is to track down the bug, the faster it is solved. 
- [ ] Feature Request. Start by telling us what problem you’re trying to solve.  
    Often a solution already exists! Don’t send pull requests to implement new features without 
    first getting our support. Sometimes we leave features out on purpose to keep the project small.

### Issue description

I'm not sure what does that error message means. I have followed the error thread which laid me to the following line 

    Assert.hasText(className, "No class name given"); // at org.elasticsearch.hadoop.util.ObjectUtils.instantiate(ObjectUtils.java:32)

### Steps to reproduce

Code:

```scala
scala> data
// res0: org.apache.spark.sql.DataFrame = [doc_id: bigint, eventDate: timestamp, marketObjectId: bigint, eventType: string, merchantId: bigint, userId: bigint]

scala> data.printSchema
// root
//  |-- doc_id: long (nullable = false)
//  |-- eventDate: timestamp (nullable = true)
//  |-- marketObjectId: long (nullable = true)
//  |-- eventType: string (nullable = true)
//  |-- merchantId: long (nullable = true)
//  |-- userId: long (nullable = true)

val config: scala.collection.mutable.Map[String, String] =
      scala.collection.mutable.Map(
        "pushdown" -> "true",
        "es.nodes" -> "localhost:9200", // params.esHost
        "es.mapping.id" -> "doc_id"
)

data.write.format("org.elasticsearch.spark.sql")
      .mode(SaveMode.Overwrite)
      .options(config)
      .save("clicks/event") 
```

Strack trace:

```
Exception in thread "main" org.elasticsearch.hadoop.EsHadoopIllegalArgumentException: No class name given
    	at org.elasticsearch.hadoop.util.Assert.hasText(Assert.java:30)
    	at org.elasticsearch.hadoop.util.ObjectUtils.instantiate(ObjectUtils.java:32)
    	at org.elasticsearch.hadoop.util.ObjectUtils.instantiate(ObjectUtils.java:52)
    	at org.elasticsearch.hadoop.util.ObjectUtils.instantiate(ObjectUtils.java:48)
    	at org.elasticsearch.hadoop.serialization.bulk.AbstractBulkFactory.initExtractorsFromSettings(AbstractBulkFactory.java:198)
    	at org.elasticsearch.hadoop.serialization.bulk.AbstractBulkFactory.<init>(AbstractBulkFactory.java:174)
    	at org.elasticsearch.hadoop.serialization.bulk.IndexBulkFactory.<init>(IndexBulkFactory.java:27)
    	at org.elasticsearch.hadoop.serialization.bulk.BulkCommands.create(BulkCommands.java:39)
    	at org.elasticsearch.hadoop.rest.RestRepository.lazyInitWriting(RestRepository.java:130)
    	at org.elasticsearch.hadoop.rest.RestRepository.writeProcessedToIndex(RestRepository.java:174)
    	at org.elasticsearch.hadoop.rest.RestRepository.delete(RestRepository.java:549)
    	at org.elasticsearch.spark.sql.ElasticsearchRelation.insert(DefaultSource.scala:481)
    	at org.elasticsearch.spark.sql.DefaultSource.createRelation(DefaultSource.scala:76)
    	at org.apache.spark.sql.execution.datasources.ResolvedDataSource$.apply(ResolvedDataSource.scala:222)
    	at org.apache.spark.sql.DataFrameWriter.save(DataFrameWriter.scala:148)
    	at org.apache.spark.sql.DataFrameWriter.save(DataFrameWriter.scala:139)
        ...
``` 

### Version Info

OS:         :  Ubuntu 14.04
JVM         :  1.8.0_101
Hadoop/Spark:  Spark 1.6.2
ES-Hadoop   :  elasticsearch-spark_2.10 v. 5.0.0.BUILD-SNAPSHOT
ES          :  5.0.0-alpha4

### Additional Info

The project is assembled with maven (mvn assembly:assembly) to create an uber-jar so all the dependencies are available.
_____

On the other hand, the reading data from elasticsearch works perfectly :  

```scala
sqlContext.read.format("org.elasticsearch.spark.sql")
              .options(config).load("clicks/event")
```
_____

I have doubt concerning the  `SaveMode.Overwrite` options. I have deleted the index that I wish to overwrite and I have tried to write the same data, that also works perfectly.

**Comment by jbaiera** [comment]
The issue here is that when SaveMode.Overwrite is specified, the connector deletes data from the current index before executing the job to insert the new data. During this deletion process, it instantiates a client object to interact with the server. The client object sees that `es.mapping.id` is set and attempts to load up an extractor object that would normally extract the IDs from documents. This extractor isn't needed in this phase because all we're doing is deleting data. The correct configuration is added to the settings right after the deletion occurs. The client only goes looking for this extractor in the settings because the `es.mapping.id` is set. If that were not configured, it would happily go about it's task. This is a strange edge case with the REST client we use and is definitely a bug.

**Comment by eliasah** [comment]
Thanks @jbaiera ! Should I pull and recompile from sources to get the fix, or is it already on maven with the build-snapshot ?

**Comment by jbaiera** [comment]
@eliasah The change should be available already in the new v2.4.0 that was released today.

**Comment by eliasah** [comment]
Sounds good to me. Thanks !
